### PR TITLE
#208 FilteringParserDelegate match count support

### PR DESCRIPTION
**Pull request by rfoltyns**
First approach to handle _matchCount. Given that `FilteringParserDelegate` is a non-trivial one - it's worth do have a double-check on this one.

### CODE REVIEW

File: src/main/java/com/fasterxml/jackson/core/filter/FilteringParserDelegate.java [hunk 1]
```diff
@@ -378,7 +378,7 @@ public JsonToken nextToken() throws IOException
                 boolean returnEnd = _headContext.isStartHandled();
                 f = _headContext.getFilter();
                 if ((f != null) && (f != TokenFilter.INCLUDE_ALL)) {
-                    f.filterFinishArray();
+                    f.filterFinishObject();
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/core/filter/FilteringParserDelegate.java]: This handles both `ID_END_ARRAY` and `ID_END_OBJECT` cases so it probably needs to call different method (check `t.id()` again or duplicate)?

File: src/main/java/com/fasterxml/jackson/core/filter/FilteringParserDelegate.java [hunk 2]
```diff
@@ -418,7 +418,7 @@ public JsonToken nextToken() throws IOException
                 }
                 _itemFilter = f;
                 if (f == TokenFilter.INCLUDE_ALL) {
-                    if (_includePath) {
+                    if (_verifyAllowedMatches() && _includePath) {
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/core/filter/FilteringParserDelegate.java]: Is it possible/easy to briefly explain check logic here?

File: src/main/java/com/fasterxml/jackson/core/filter/FilteringParserDelegate.java [hunk 3]
```diff
@@ -545,7 +547,7 @@ protected final JsonToken _nextToken2() throws IOException
                     boolean returnEnd = _headContext.isStartHandled();
                     f = _headContext.getFilter();
                     if ((f != null) && (f != TokenFilter.INCLUDE_ALL)) {
-                        f.filterFinishArray();
+                        f.filterFinishObject();
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/core/filter/FilteringParserDelegate.java]: As per earlier case, applies both to arrays and objects.

File: src/main/java/com/fasterxml/jackson/core/filter/FilteringParserDelegate.java (hunk 2 above)
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/core/filter/FilteringParserDelegate.java]: Maybe switch order of checks (since boolean check is faster than call)?

**Comment by cowtowncoder** [comment]
I think it makes sense at high level; one flaw (wrt assuming it's always closing Object scope), otherwise looks good.

File: src/main/java/com/fasterxml/jackson/core/filter/FilteringParserDelegate.java (hunk 2 above)
Comment (rfoltyns) [inline src/main/java/com/fasterxml/jackson/core/filter/FilteringParserDelegate.java]: Sure. I've found it at `testNotAllowMultipleMatchesWithoutPath1` test. It's easier to explain based on the line `FilterParserDelegate`:586 - the condition is the same.
If the first match occurs, `_matchCount` has to be incremented and regardless if the `_includePath` is true or false - the rest of the logic will handle both. However, if we invert it - logic will do the job - the output will be ok, but `_matchCount` will stay at 0.

In this commit, `FilterParserDelegate`:421 it's just a copy of the other condition, but I've found it crucial for another PR that I'll publish after this one (more edge cases handled, more tests, one thing at a time)

File: src/main/java/com/fasterxml/jackson/core/filter/FilteringParserDelegate.java (hunk 1 above)
Comment (rfoltyns) [inline src/main/java/com/fasterxml/jackson/core/filter/FilteringParserDelegate.java]: Yeah, at first, I thought about it just as a cosmetic change as `com.fasterxml.jackson.core.filter.TokenFilter` have the same implementations fo both array and objects. But, if we want do it properly here, we should probably have to copy some code and call proper methods in ID_END_OBJECT and ID_END_ARRAY.

It's definitely out of scope of this commit, so I can move it to another issue if you want.

Comment (rfoltyns) [inline src/main/java/com/fasterxml/jackson/core/filter/FilteringParserDelegate.java]: I've reverted this change. I doesn't make much sense in this issue.

File: src/main/java/com/fasterxml/jackson/core/filter/FilteringParserDelegate.java (hunk 2 above)
Comment (rfoltyns) [inline src/main/java/com/fasterxml/jackson/core/filter/FilteringParserDelegate.java]: You're right, it would be faster, but it will not count properly anymore (please have a look at my previous comment)

Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/core/filter/FilteringParserDelegate.java]: Oh... so, `verifyAllowedMatches()` has a side-effect. Ok, makes sense, I should have noticed.
